index.js

```diff
@@ -1,159 +1,201 @@
-  throw new Error('ADMIN_ACCESS_TOKEN not set!');
+if (!CLIENT_ACCESS_TOKEN) {
+  throw new Error("CLIENT_ACCESS_TOKEN not set!");
 }
-if(!CLIENT_ACCESS_TOKEN) {
-  throw new Error('CLIENT_ACCESS_TOKEN not set!');
+if (!SESSION_SECRET) {
+  throw new Error("SESSION_SECRET not set!");
 }
-if(!SESSION_SECRET) {
-  throw new Error('SESSION_SECRET not set!');
-}
-
 
-function adminAuth(app, config, services) {
+function enableKeycloakOauth(app, config, services) {
   const { baseUriPath } = config.server;
   const { userService, accessService } = services;
-  
+
   passport.use(
-    new githubStrategy(
+    "keycloak",
+    new KeycloakStrategy(
       {
-        clientID: GH_CLIENT_ID,
-        clientSecret: GH_CLIENT_SECRET,
-        callbackURL: GH_CALLBACK_URL,
+        host: KC_HOST,
+        realm: KC_REALM,
+        clientID: KC_CLIENT_ID,
+        clientSecret: KC_CLIENT_SECRET,
+        callbackURL: `${baseUriPath}/api/auth/callback`,
+        authorizationURL: `${KC_HOST}/realms/${KC_REALM}/protocol/openid-connect/auth`,
+        tokenURL: `${KC_HOST}/realms/${KC_REALM}/protocol/openid-connect/token`,
+        userInfoURL: `${KC_HOST}/realms/${KC_REALM}/protocol/openid-connect/userinfo`,
       },
-      async (accessToken, refreshToken, profile, cb) => {
-  
-        var role;
-  
-        for (var index in profile.teams) {
-          if (READ_ONLY_TEAM && profile.teams[index].name === READ_ONLY_TEAM) {
-            role = "Viewer";
-          }
-          if (profile.teams[index].name === TEAM) {
-            role = "Editor";
-            break;
-          }
-        }
-  
-        if (!role) {
-          cb(null, false);
+
+      async (accessToken, refreshToken, profile, done) => {
+        console.log("profile", profile);
+
+        const user_roles = profile.roles || [];
+        const isAdmin = user_roles.some((item1) =>
+          KC_ADMIN_ROLES.some(
+            (item2) => item2.toLowerCase() === item1.toLowerCase()
+          )
+        );
+        const isEditor = user_roles.some((item1) =>
+          KC_EDITOR_ROLES.some(
+            (item2) => item2.toLowerCase() === item1.toLowerCase()
+          )
+        );
+        const isViewer = user_roles.some((item1) =>
+          KC_VIEWER_ROLES.some(
+            (item2) => item2.toLowerCase() === item1.toLowerCase()
+          )
+        );
+
+        if (isAdmin) {
+          var role = "Admin";
+        } else if (isEditor) {
+          var role = "Editor";
+        } else if (isViewer) {
+          var role = "Viewer";
         } else {
-          const email  = profile.emails[0].value;
-          const user = await userService.loginUserWithoutPassword(email, true);
-          await accessService.setUserRootRole(user.id, role);
-          cb(null, user);
-          }
-      },
-    ),
+          // Not authorized
+          console.log("Not authorized");
+          done(null, false);
+          return;
+        }
+        console.log("user role", role);
```

Comment:
`var` shouldn't be used, use `const` or `let` instead, since `isAdmin`, `isEditor` and `isViewer` only used to compute `role`, can extract the whole logic into a function like

```javascript
const getRole = (user_roles) => {
        if ((user_roles.some((item1) =>
          KC_ADMIN_ROLES.some(
            (item2) => item2.toLowerCase() === item1.toLowerCase()
          )
        )) {
          return "Admin";
        }
        if (user_roles.some((item1) =>
          KC_EDITOR_ROLES.some(
            (item2) => item2.toLowerCase() === item1.toLowerCase()
          )
        )) {
           return "Editor";
        }
        if (user_roles.some((item1) =>
          KC_VIEWER_ROLES.some(
            (item2) => item2.toLowerCase() === item1.toLowerCase()
          )
        )) {
          return "Viewer"
        }
        return null;
}

```

Reply:
Thank you, I applied your suggestion.